app/controllers/application_controller.rb

```diff
@@ -5,6 +5,86 @@ class ApplicationController < Sinatra::Base
   configure do
     set :public_folder, 'public'
     set :views, 'app/views'
+    enable :sessions
+    set :session_secret, "78y482eh2o"
+  end
+
+ get "/" do
+ 	if logged_in?
+      redirect "/tweets"
+    else
+    erb :index
+    end
+  end
+
+  get "/signup" do
+  	if logged_in?
+      redirect "/tweets"
+    else
+      erb :signup
+    end
+  end
+
+  post "/signup" do
+    if params[:username] == "" || params[:password] == "" || params[:email] == ""
+      redirect "/signup"
+      else
+      user = User.create(username: params[:username], email: params[:email], password: params[:password] )
+      session[:user_id] = user.id
+      redirect "/tweets"
+    end
+  end
+
+
+  get "/login" do
+  	if logged_in?
+      redirect "/tweets"
+    else
+      erb :login
+    end
+  end
+
+  post "/login" do
+    if params[:username] == "" || params[:password] == ""
+      redirect "/failure"
+     else
+      user = User.find_by(username: params[:username] )
+      if user && user.authenticate(params[:password])
+        session[:user_id] = user.id
+        redirect "/tweets"
+        else
+        redirect "/failure"
+      end
+    end
+  end
+
+
+  get "/failure" do
+    erb :failure
+  end
+
+  get "/logout" do
+  	if !logged_in?
+      redirect "/"
+    else
+    session.clear
+    redirect "/login"
+    end
+  end
+
+  get '/users/:slug' do
+    @user = User.find_by_slug(params[:slug])
+    erb :'/users/show'
+  end
+
+  helpers do
+    def logged_in?
+      !!session[:user_id]
```

Comment:
It's better to rely on the successful call to `current_user` than only introspecting on the `session[:id]` value. In theory, a user might have an invalid `id` in their session which would make `logged_in?` return true even though that user would not be findable later. Also, because `current_user` is memoized, it makes sense to call it as early in the request cycle as possible, which generally is the first call `logged_in?`.

```ruby
def logged_in?
  !!current_user
end
```